Precache rollup values

kbrock opened the pull request:
reference:
- https://github.com/ManageIQ/manageiq/issues/22593

For Metrics rollup, we fetch the `container_image` from each of the `container` objects.
This change preloads the parent `container_image` value into the `containers` so we avoid the N+1.

TODO: I need to followup to find the test case to ensure I'm fixing the issue at hand

before
======

```ruby
container_image = ContainerImage.first
containers = container_image.vim_performance_state(timestamp).containers
containers.map(&:container_image) # N+1
```

after
=====

```ruby
container_image = ContainerImage.first
containers = container_image.vim_performance_state(timestamp).containers
containers.map(&:container_image) # ==> array of container_image object
```

Fryguy commented on lib/miq_preloader.rb:
```diff
@@ -1,21 +1,53 @@
+  #
+  # example:
+  #   vms = Vm.where(:ems_id => emses.id)
+  #   preload_reverse(emses, :vms, vms)
```
```suggestion
  #   preload(emses, :vms, vms)
```

Fryguy commented on lib/miq_preloader.rb:
```diff
@@ -1,21 +1,53 @@
+  # also of interest:
+  # reverse_association = record.class.reflect_on_association(association).inverse_of.name
+  # ra = value.association(reverse_association) # vms.ems
+
```
```suggestion
```

kbrock replied:
lol. to be fair. this is not a double space :)

Fryguy replied:
HAHA - interesting timing to our discussion (also I considered this a one off mistake :) )

Fryguy commented on lib/miq_preloader.rb:
```diff
@@ -1,21 +1,53 @@
+  #
+  # example:
+  #   vms = Vm.where(:ems_id => ems.id)
+  #   preload_reverse(ems, :vms, vms)
```
Is this symbol correct?  I think this is supposed to be:

```suggestion
  #   preload_reverse(ems, :ems, vms)
```

Fryguy replied:
Otherwise I don't understand this symbol

kbrock commented:
update:
- fixed comments
- fixed new preload call (I had swapped the arguments)

Fryguy commented on spec/lib/miq_preloader_spec.rb:
```diff
@@ -4,29 +4,58 @@
+      expect { vms.map(&:ext_management_system) }.to make_database_queries
+
+      vms.each(&:reload)
+      expect { MiqPreloader.preload_reverse(ems, :vms, vms) }.not_to make_database_queries
```
Oh but now I'm seeing :ems in the test - this is confusing.  I think I know an interface that would make this less complicated, but I need to discuss.

kbrock commented:
Moving the guts to https://github.com/ManageIQ/manageiq/pull/22589

Keeping this PR for comments

Fryguy commented on lib/miq_preloader.rb:
```diff
@@ -1,4 +1,26 @@
+  # If you want to preload an association on multiple records
+  # or want to only load a subset of an association
+  #
+  # example:
```
```suggestion
  # @example Preloading vms on a set of emses
```

Fryguy commented on lib/miq_preloader.rb:
```diff
@@ -1,4 +1,26 @@
+  # or want to only load a subset of an association
+  #
+  # example:
+  #   vms = Vm.where(:ems_id => emses.id)
+  #   preload(emses, :vms, vms)
+  #   emses.map { |ems| ems.vms } # cached - no queries
+  #   vms.first.ems # cached - the reversed association is cached
```
```suggestion
  #   vm_scope = Vm.where(:ems_id => emses.id)
  #   preload(emses, :vms, vm_scope)
  #   emses.map { |ems| ems.vms } # cached - no queries
  #   vms.first.ems # cached - the reversed association is cached
```

Fryguy commented on lib/miq_preloader.rb:
```diff
@@ -1,4 +1,26 @@
+  #   emses.map { |ems| ems.vms } # cached - no queries
+  #   vms.first.ems # cached - the reversed association is cached
+  #
+  #   Do the folowing to programitically determine the reverse association name (e.g.: :ems)
```
```suggestion
  #   Do the following to programmatically determine the reverse association name (e.g.: :ems)
```

Fryguy commented on lib/miq_preloader.rb:
```diff
@@ -1,4 +1,26 @@
+  #   reverse_association_name = record.class.reflect_on_association(association).inverse_of.name # :ems
+  #   reverse_association = result.association(reverse_association_name) # vms.first.ems
+  #
+  # @param record [relation|ActiveRecord::Base|Array[ActiveRecordBase]]
```
```suggestion
  # @param record [relation|ActiveRecord::Base|Array[ActiveRecord::Base]]
```

Fryguy commented on lib/miq_preloader.rb:
```diff
@@ -1,4 +1,26 @@
+  #
+  # @param record [relation|ActiveRecord::Base|Array[ActiveRecordBase]]
+  # @param association [Symbol|Hash|Array] name of the association(s)
+  # @param preload_scope [Nil|relation] relation of the records.
+  #        for all but one case, default behavior is to use the association
+  #        alternatively a scope can be used.
+  #        But currently an array does not work
```
```suggestion
  # @param preload_scope [Nil|relation] Relation of the records to be used for preload
  #        For all but one case, default behavior is to use the association.
  #        Alternatively a scope can be used.
  #        Currently an array does not work
```

Fryguy commented on lib/miq_preloader.rb:
```diff
@@ -10,12 +32,15 @@ def self.preload(records, associations, preload_scope = nil)
   #   preload_and_map(orchestration_stack.subtree, :direct_vms)
+  #
+  # @param records [ActiveRecord::Base, Array<ActiveRecord::Base>, Object, Array<Object>]
+  # @param association_name [Symbol] an association on records
```
```suggestion
  # @param association [Symbol] name of the association
```

Fryguy commented on lib/miq_preloader.rb:
```diff
@@ -10,12 +32,15 @@ def self.preload(records, associations, preload_scope = nil)
 
   # @param records [ActiveRecord::Base, Array<ActiveRecord::Base>, Object, Array<Object>]
-  # @param association [String] an association on records
+  # @param association_name [Symbol] an association on records
```
```suggestion
  # @param association_name [Symbol] name of the association
```

Fryguy commented on spec/lib/miq_preloader_spec.rb:
```diff
@@ -4,22 +4,35 @@
+
       expect(ems.vms).to be_loaded
       expect { preload(ems, :vms) }.to_not make_database_queries
+      expect { ems.vms.size }.not_to make_database_queries
```
Keeping consistent with the line before it.  (although it's not_to on line 5 also, so 🤷 )

```suggestion
      expect { ems.vms.size }.to_not make_database_queries
```

kbrock replied:
I `s/to_not/not_to/g` to make consistent.

Other places said not to split the participle (to verb)

Fryguy commented:
@kbrock Now that we decided to switch to using the original reload and we got rid of the "extra" method, is there anything substantive in this PR anymore?  Seems it's just yard docs and specs.  If that't the case can you update the OP and perhaps the labels?

kbrock commented:
update:

- `s/to_not/not_to` (specs only)
- changed VimPerformanceStates to use scopes
- changed VimPerformanceState#capture (spec only)
- VimPerformanceState#vim_performance_state_association preloads forward/reverse association

update:

- fix white space issues

kbrock commented:
update:
- took one commit from #22589
  - requires that we use scopes

Fryguy commented:
Backported to `petrosian` in commit 3987deafd36d4aae5064660e544a0af328faaf33.

```text
commit 3987deafd36d4aae5064660e544a0af328faaf33
Author: Jason Frey <fryguy9@gmail.com>
Date:   Wed Jul 5 15:55:18 2023 -0400

    Merge pull request #22594 from kbrock/precache_rollup_values
    
    Precache rollup values
    
    (cherry picked from commit f28d10779b15316e27578175c07145a3b35469bf)
```
